lua/lean/rpc.lua

```diff
@@ -18,50 +17,63 @@ local control = require'plenary.async.control'
 local Session = {}
 Session.__index = Session
 
+---@param client any vim.lsp.client object
 ---@param bufnr number
 ---@param uri string
 ---@return Session
-function Session:new(bufnr, uri)
+function Session:new(client, bufnr, uri)
   self = setmetatable({
-    bufnr = bufnr,
+    client = client,
     uri = uri,
     session_id = nil,
     connected = nil,
     connect_err = nil,
-    closed = false,
     on_connected = control.Condvar.new(),
     to_release = {},
     release_timer = nil,
   }, self)
   self.keepalive_timer = vim.loop.new_timer()
   self.keepalive_timer:start(20000, 20000, vim.schedule_wrap(function()
-    if self.session_id ~= nil then
-      vim.lsp.buf_notify(self.bufnr, '$/lean/rpc/keepAlive', {
+    if not self:is_closed() and self.session_id ~= nil then
+      self.client.notify('$/lean/rpc/keepAlive', {
         uri = self.uri,
         sessionId = self.session_id,
       })
     end
   end))
   -- Terminate RPC session when document is closed.
   vim.api.nvim_buf_attach(bufnr, false, {
-    on_reload = function() self.closed = true end,
-    on_detach = function() self.closed = true end,
+    on_reload = function() self:close_without_releasing() end,
+    on_detach = function() self:close_without_releasing() end,
   })
```

Comment:
ISTM that the fix to #189 consists of these lines? I wanted to ask why using the client directly is necessary when `vim.lsp.buf_notify` is designed to avoid the need for this, and in particular it seems very unlikely that a different client will have a `$/lean/rpc/keepAlive` handler. Seems a tad more readable to just pass in the buffer, but this is minor so feel free to go ahead if you're more comfortable with this.

Reply:
I think the main fix is checking `is_stopped()` to figure out if the old server is stopped, the handlers here are only called when the buffer is reloaded.  This doesn't happen with `:LspRestart` iirc.  I've also fixed all the other places where the session is closed, and stopped the keepalive timer there as well.

> I wanted to ask why using the client directly is necessary [...]

It's more for clarity.  We don't want to send the requests to any server supporting the request, but they need to be sent to exactly to the server we've connected to.  Other servers (or even a restarted server) won't know about the session id, the RPC references, etc.  Using the client instead of the buffer number expresses this intention.

Reply:
OK, sounds good to me.